### DISCUSSION TITLE
Tracked notifyOnChangeProps on separate components

### DISCUSSION THREAD

**Discussion by talberkoMelio** [TanStack/query#8487] [Q&A]
I use the same query across different components.
One component uses some props from the query (data, isLoading etc), and another one only triggers a fetch for prefetching without using it.

By using the "tracked" notifyOnChangeProps, will the component which only prefetches the query be rerendered when the data is changed?

Because when debugging it I do see it still rerenders.

Would appreciate your feedback

**Comment by TkDodo** [comment]
How is the prefetch triggered?

**Reply by talberkoMelio** [reply]
@TkDodo in the following example I have the same query used in two places - parent and a child.
The parent doesnt track any return value from the query, the child does track.

In the first example below there are 2 times "Parent component render" log, even there is no props in use in the Parent component. "Child component render"  also logged twice (which makes sense).

```
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      notifyOnChangeProps: "tracked",
    },
  },
});

export default function App() {
  return (
    <QueryClientProvider client={queryClient}>
      <Parent />
    </QueryClientProvider>
  );
}

function Child() {
  const { isLoading, data } = useQuery("repoData", () =>
    axios
      .get("https://api.github.com/repos/tannerlinsley/react-query")
      .then((res) => res.data)
  );

  console.log("Child component render");

  return <div>hi</div>;
}

function Parent() {
  useQuery("repoData", () =>
    axios
      .get("https://api.github.com/repos/tannerlinsley/react-query")
      .then((res) => res.data)
  );

  console.log("Parent component render");

  return <Child />;
}
```

If I will change the Parent component to the code below, there will be one time "Parent component render" log, which makes sense, and "Child component render" will be twice which is expected.

```
function Parent() {
  useQuery(
    "repoData",
    () =>
      axios
        .get("https://api.github.com/repos/tannerlinsley/react-query")
        .then((res) => res.data),
    {
      notifyOnChangeProps: [],
    }
  );

  console.log("Parent component render");

  return <Child />;
}
```

I am not sure what is the expected behavior of the "tracked" value, but it seems it doesnt track it per observer. Any idea?

**Reply by TkDodo** [reply]
if you call `useQuery` without using any property, we track all properties as a safety measure. If you want prefetching,  we have a `usePrefetchQuery` hook for that which doesn’t create a subscription at all.

**Reply by talberkoMelio** [reply]
@TkDodo Interesting. I think it worth mention it in the docs. 
And if I call the useQuery in the following way:
const query = useQuery(...);

in this case, also all props will be tracked, or none of them unless explicitly accessed ("query.isLoading" for example)?

(BTW I use version 3 so I dont have the prefetch hook)

**Reply by TkDodo** [reply]
All of them, unless you explicitly access them.

**Reply by talberkoMelio** [reply]
@TkDodo not sure I get it.
Shouldn't it be:
"None of them, unless you explicitly access them?"

If I understand your answer, I should always access the relevant members like this to use the "tracked" option like this:
const { isLoading, data } = useQuery()

Otherwise, when using the following ways it will create subscription for all members:
const query = useQuery() OR
useQuery()

Am I right?

**Reply by talberkoMelio** [reply]
@TkDodo also from your blog it looks that when accessing this way it will not track, which I understand it as "None of them, unless you explicitly access them" and not "All of them, unless you explicitly access them.". 

I played with a sandbox and seems like using it this does track all props, so I am confused.

May you please elaborate?

<img width="1004" alt="image" src="https://github.com/user-attachments/assets/b80448d0-c8b7-4cf1-91b5-8b6c33764ffc" />

**Reply by TkDodo** [reply]
> Shouldn't it be:
"None of them, unless you explicitly access them?"

no. if you don’t access any properties, we will track all properties:

```
const query = useQuery(...)

console.log(query)

return null
```

this will re-render on every property change, as “never re-rendering anything” is almost never what you want.

> I should always access the relevant members like this to use the "tracked" option like this:
const { isLoading, data } = useQuery()

You can do that, or you can do:

```
const query = useQuery()

return <div>{query.data}, {query.isLoading}</div>
```

destructuring or “normal” access with `query.something` is equivalent.

> also from your blog it looks that when accessing this way it will not track,

wrong; `queryInfo.data` is accessed during render, so data will be tracked.

> I played with a sandbox and seems like using it this does track all props, so I am confused.

then there’s something else wrong in your sandbox 🤷‍♂️.

**Reply by talberkoMelio** [reply]
@TkDodo got it now. In your first example, even when there will be no console log of the query the component will rerender when every prop will change, correct?
```
const query = useQuery(...)

// there will be rerender here when any of the query props changes

return null
```

Is there a way to change it so if not using its props it wont rerender?
If a query is disabled, will it still have subscriptions for its members?

In our (huge) codebase there is a wrapper of the query which contains CRUD queries and mutations together, so basically when you use the wrapper you will get the query, createMutation, deleteMutation, updateMutation, which means a subscription for all when you probably just need the query.
Any suggestion here?

**Reply by TkDodo** [reply]
> In your first example, even when there will be no console log of the query the component will rerender when every prop will change, correct?

yes

> Is there a way to change it so if not using its props it wont rerender?

pass `notifyOnChangeProps: []` to opt out of all re-renders.

> If a query is disabled, will it still have subscriptions for its members?

yes. disabled just means “this observer won’t fetch data on its own”.

> Any suggestion here?

Don’t do that 😅 . Creating query subscriptions via observers when you only need a mutation is a bad idea. I don’t recall ever recommending such ”monster hooks”.
